gapis/resolve/dependencygraph2/builder.go

```diff
@@ -0,0 +1,303 @@
+func (b *dependencyGraphBuilder) Graph() DependencyGraph {
+	return b.graph
+}
+
+// Build a new DependencyGraphBuilder.
+func NewDependencyGraphBuilder(ctx context.Context, config DependencyGraphConfig,
+	c *capture.Capture, initialCmds []api.Cmd, initialCmdRanges interval.U64RangeList) DependencyGraphBuilder {
+	graph := newDependencyGraph(ctx, config, c, initialCmds, initialCmdRanges)
+	builder := &dependencyGraphBuilder{
+		graph: graph,
+		currentDependencyExists: make([]bool, graph.NumNodes()),
+		currentCmdID:            api.CmdNoID,
+		currentNodeID:           NodeNoID,
+		fragmentWrites:          make(map[api.RefID]map[api.Fragment]NodeID),
+		memoryWrites:            make(map[memory.PoolID]*memoryWriteList),
+		openForwardDependencies: make(map[interface{}]NodeID),
+	}
+	return builder
+}
+
+// BeginCmd is called by the API emulator at the beginning of each API call
+func (b *dependencyGraphBuilder) OnBeginCmd(ctx context.Context, cmdID api.CmdID, cmd api.Cmd) {
+	debug(ctx, "OnBeginCmd [%d] %v", cmdID, cmd)
+	b.currentCmdID = cmdID
+	b.currentCmd = cmd
+	b.currentNodeID = b.graph.GetNodeID(CmdNode{api.SubCmdIdx{uint64(cmdID)}})
+}
+
+// EndCmd is called by the API emulator at the end of each API call
+func (b *dependencyGraphBuilder) OnEndCmd(ctx context.Context, cmdID api.CmdID, cmd api.Cmd) {
+	debug(ctx, "OnEndCmd [%d] %v", cmdID, cmd)
+	if len(b.currentDependencies) > 0 {
+		sortedDeps := make([]NodeID, len(b.currentDependencies), len(b.currentDependencies)+b.currentCreateCount)
+		copy(sortedDeps, b.currentDependencies)
+		b.currentDependencies = b.currentDependencies[:0]
+		sort.Slice(sortedDeps, func(i, j int) bool { return sortedDeps[i] < sortedDeps[j] })
+		for _, tgt := range sortedDeps {
+			b.currentDependencyExists[tgt] = false
+		}
+		b.graph.setDependencies(b.currentNodeID, sortedDeps)
+		b.currentCmdID = api.CmdNoID
+		b.currentCmd = nil
+		b.currentCreateCount = 0
+		b.currentNodeID = NodeNoID
+	}
+}
+
+func (b *dependencyGraphBuilder) OnGet(ctx context.Context, owner api.Reference, frag api.Fragment, valueRef api.Reference) {
+	debug(ctx, "  OnGet (%T %d)%v : %d", owner, owner.RefID(), frag, valueRef.RefID())
+	readEffect := ReadFragmentEffect{b.currentNodeID, frag}
+	writes, ok := b.fragmentWrites[owner.RefID()]
+	if !ok {
+		return
+	}
+	if _, ok := frag.(api.CompleteFragment); ok {
+		for writeFrag, writeNode := range writes {
+			writeEffect := WriteFragmentEffect{writeNode, writeFrag}
+			b.addDependency(writeEffect, readEffect)
+		}
+	} else if writeNode, ok := writes[frag]; ok {
+		writeEffect := WriteFragmentEffect{writeNode, frag}
+		b.addDependency(writeEffect, readEffect)
+	}
+}
+
+func (b *dependencyGraphBuilder) OnSet(ctx context.Context, owner api.Reference, frag api.Fragment, oldValueRef api.Reference, newValueRef api.Reference) {
+	debug(ctx, "  OnSet (%T %d)%v : %d → %d", owner, owner.RefID(), frag, oldValueRef.RefID(), newValueRef.RefID())
+	if _, ok := frag.(api.CompleteFragment); ok {
```

Comment:
Actually I don't find any case of OnSet(..., api.CompleteFragment{}...). Feel like this is actually not a valid case, do we really want to handle such cases? Or we should panic or emit error?

Reply:
You're right that `CompleteFragment` is not currently used with `OnSet`, but this is a perfectly valid case. I think it is better to keep this case, so that `OnSet` can correctly handle all types of `Fragments`.

This was used previously when non-reference structs contained their data directly, rather than through a `data` pointer. At that point, assigning to a struct was the same as assigning to all of it's fields, which was represented by `OnSet` with a `CompleteFragment`.

I can also imagine other scenarios that might use `CompleteFragment` with `OnSet`: e.g., `Clear` method for maps.

Reply:
RE: `oldValueRef` and `newValueRef`, these are helpful for debugging. They can also be used to reconstruct a human readable path to the fragments (to get something like `Buffers[0x1234].Info.Size` instead of `0x6789.Size`). However, maintaining enough data to reconstruct these human readable paths is expensive, and unnecessary for DCE, so I have not included that code in this PR.